src/main/java/ui/UI.java

```diff
@@ -0,0 +1,186 @@
+     */
+    public String getUserInput() {
+        try {
+            return scanner.nextLine().trim();
+        } catch (Exception e) {
+            printResponse("Error reading input: " + e.getMessage());
+            return ""; // Return an empty string
+        }
+    }
+
+    /**
+     * Prints a greeting message to the console.
+     */
+    public void sayHi() {
+        printResponse(Messages.WELCOME);
+    }
+
+    /**
+     * Prints a farewell message to the console.
+     */
+    public void cleanupAndExit() {
+        printResponse("Adios, amigo!");
+        scanner.close();
+    }
+
+    /**
+     * Prints table displaying student information.
+     * The table includes student IDs, GPAs, and preference rankings.
+     *
+     * @param studentList ArrayList of Student objects to be printed
+     * @throws SEPEmptyException if the student list is null or empty.
+     */
+    public void printStudentList(ArrayList<Student> studentList) throws SEPEmptyException {
+        if (studentList == null || studentList.isEmpty()) {
+            throw SEPEmptyException.rejectEmptyStudentList();
+        }
+
+        AsciiTable at = new AsciiTable();
+        at.addRule();
+        at.addRow("Student", "GPA", "Preference Rankings");
+        at.addRule();
+
+        for (Student s : studentList) {
+            // Using Java Streams to create a comma-separated preference rankings string
+            String uniPrefsString = s.getUniPreferences().stream()
+                    .map(String::valueOf) // Convert each Integer to String
+                    .collect(Collectors.joining(",")); // Join them with commas
+
+            // Use the uniPrefsString for adding to the table
+            at.addRow(s.getId(), s.getGpa(), uniPrefsString);
+        }
+        at.addRule();
+
+        at.setTextAlignment(TextAlignment.CENTER);
+        at.setPaddingLeft(3);
+        at.setPaddingRight(3);
+
+        at.getRenderer().setCWC(new CWC_LongestLine());
+        printResponse("Here is the list:\n" + at.render());
+    }
+
+    /**
+     * Prints a buffering message when the algorithm is running.
+     * Upon algorithm completion, replaces the buffering message with a
+     * completed message.
+     */
+    public void printAllocatingMessage() {
+        System.out.println(HORIZONTAL_LINE);
+        try {
+            System.out.print("\rLoading");
+            while (!Thread.currentThread().isInterrupted()) { // Keep printing until interrupted
+                for (int i = 0; i <= 3; i++) {
+                    System.out.print("\rLoading" + ".".repeat(i));
+                    Thread.sleep(500); // 500 milliseconds
+                }
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt(); // Restore the interrupted status
+        } finally {
+            System.out.println("\r" + Messages.ALLOCATE_COMPLETE);
+            System.out.println(HORIZONTAL_LINE);
+        }
+    }
```

Comment:
You could consider converting some numbers and strings into named constant to give it a meaning. It will be easier for future usage or change as well.